include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h

```diff
@@ -412,68 +412,71 @@ using __repacked_tuple_t = typename __repacked_tuple<T>::type;
 template <typename _ContainerOrIterable>
 using __value_t = typename __internal::__memobj_traits<_ContainerOrIterable>::value_type;
 
-//-----------------------------------------------------------------------
-// future and helper classes for async pattern/algorithm
-//-----------------------------------------------------------------------
+template <typename _Event>
+void __wait_event(_Event);
 
-// TODO: towards higher abstraction and generic future. implementation specific sycl::event should be hidden
-struct __future_base
+void
+__wait_event(sycl::event __e)
 {
-    sycl::event __my_event;
-
-    __future_base() = default;
-    __future_base(sycl::event __e) : __my_event(__e) {}
-    void
-    wait()
-    {
 #if !ONEDPL_ALLOW_DEFERRED_WAITING
-        __my_event.wait_and_throw();
+    __e.wait_and_throw();
 #endif
-    }
-    operator sycl::event() const { return __my_event; }
-};
+}
 
 template <typename _T>
-class __future : public __future_base
+constexpr auto
+__get_value(sycl::buffer<_T>& __buf)
 {
-    ::std::size_t __result_idx;
-    sycl::buffer<_T> __data;
-
-  public:
-    __future(sycl::event __e, size_t __o, sycl::buffer<_T> __b)
-        : __par_backend_hetero::__future_base(__e), __result_idx(__o), __data(__b)
-    {
-    }
+    //according to a contract, returned value is one-element sycl::buffer
+    return __buf.template get_access<access_mode::read>()[0];
+}
 
-    _T
-    get()
-    {
-        return __data.template get_access<access_mode::read>()[__result_idx];
-    }
-    template <class _Tp, class _Enable>
-    friend class oneapi::dpl::__internal::__future;
-};
+template <typename _T>
+constexpr auto
+__get_value(_T& __val)
+{
+    return __val;
+}
 
-template <>
-class __future<void> : public __future_base
+//A contract for future class: <sycl::event or other event, a value or sycl::buffers...>
+//Impl details: inheretance (private) instead of aggregation for enabling the empty base optimization.
+template <typename _Event, typename... Args>
```

Comment:
I believe `Args` should also be with underscore at the beginning

Reply:
done